README
%% /u/sy/beebe/tex/bibclean/bibclean-2.16/README, Tue Jan  1 08:59:06 2013
%% Edit by Nelson H. F. Beebe <beebe@math.utah.edu>
%% /u/sy/beebe/tex/bibclean/2-11/README, Mon Oct  2 10:01:31 1995
%% Edit by Nelson H. F. Beebe <beebe@sunrise>

========================================================================

==========
Jump start
==========

As with most GNUware, you can build, test, and install this program on
most UNIX systems by these simple steps

csh et amici:
	setenv CC ...your favorite C or C++ compiler...
	./configure && make all check install

sh et amici:
	CC=...your favorite C or C++ compiler...
	export CC
	./configure && make all check install

If you don't set the CC environment variable, then gcc (or cc, if gcc
is not available) will be assumed.

If you wish to undo a "make install", just do "make uninstall"; this
will remove any files in system directories put there by "make
install".

See below for further details, and for instructions for non-UNIX systems.
	

============
Introduction
============

This directory contains bibclean, a BibTeX prettyprinter, portability
verifier, and syntax checker.  It can be used to find errors in .bib
files, as well as to standardize their format for readability and
editing convenience.  It can also be used to convert Scribe-format
bibliographies to BibTeX form.

Binary executables for IBM PC DOS, DEC Alpha OpenVMS, DEC VAX VMS, and
Intel x86 Linux may be included in the distribution.

If you do not require either the IBM PC DOS or LINUX, or the DEC VMS
(Alpha and VAX) versions, then you can save about 2.5MB of disk space
by deleting the ibmpc and vms subdirectories.

The default pattern matching in bibclean.c is selected by
HAVE_PATTERNS; with it, no regular-expression library support is
needed.  Should you wish to compile with regular-expression support
instead of the HAVE_PATTERNS code, and your system does not have
compile()/step (HAVE_REGEXP), or re_comp()/re_exec() (HAVE_RECOMP),
you may be able to use the regex-?-??.tar.Z distribution from the Free
Software Foundation, available on prep.ai.mit.edu in /pub/gnu.

In most cases, the HAVE_PATTERNS code is recommended, since it will
give identical results across all machines.  I was prompted to write
it after discovering that there was considerable variety in the
regular expression library codes that resulted in different matching
on different machines, a most unsatisfactory situation.

Please report all problems, suggestions, and comments to the author:

	Nelson H. F. Beebe
	University of Utah
	Department of Mathematics, 110 LCB
	155 S 1400 E RM 233
	Salt Lake City, UT 84112-0090
	USA
	Tel: +1 801 581 5254
	FAX: +1 801 581 4148
	Email: beebe@math.utah.edu, beebe@acm.org, beebe@computer.org (Internet)
	WWW URL: http://www.math.utah.edu/~beebe/


============
Installation
============

Starting with version 2.10.1, bibclean has been adapted to use the GNU
autoconf automatic configuration system for UNIX installations.

GNU autoconf is run at the author's site to produce the configure
script from configure.in.

The configure script is run at each installer's UNIX site to produce
Makefile from Makefile.in, and config.h from config.hin.  The
configure script is a large (1800+ lines) Bourne shell program that
investigates various aspects of the local C implementation, and
records its conclusions in config.h.  Interestingly, its probes
uncovered a bug in one compiler: lcc 3.4b on Sun Solaris 2.x has an
incorrect definition of toupper() in its ctype.h!

autoconf, at least at the current 2.4 version, is not as C++-aware as
it should be.  The Makefile must carry out minor edits of the
configure script to get it to even work with C++ compilers.  The small
test programs run by configure to determine the existence of assorted
Standard C library functions all lead to incorrect conclusions for
config.h, because they intentionally contain function prototypes with
different argument types.  Since C++ functions are compiled into
external names that encode the function and argument types, along with
the function name, these prototypes produce references to non-existent
functions, causing program linking to fail.  Fortunately, I've been
able to fix this problem too with additional automatic edits, all
carried out by "make configure".  

Should you do a "make maintainer-clean" [NOT recommended, except at the
author's site], the configure script will be deleted, and you will
need recent versions of both GNU m4 and autoconf correctly installed
to reconstruct things, which can be done this way:

	autoconf	# Regenerate unedited configure
	./configure	# Regenerate config.h and Makefile
	rm configure	# delete configure
	make configure	# Regenerate edited configure

For convenience and safety, the distribution includes a subdirectory
named save that contains read-only copies of the files Makefile,
config.h, and configure created by autoconf and "make configure".
This will allow recovery from a lost or damaged configure file.

Suitable hand-crafted config.h files are provided for non-UNIX
systems, and in the unlikely event of a failure of the configure
script on a UNIX system, config.h can be manually produced from a copy
of config.hin with a few minutes' editing work. If you do this,
remember to save a copy of your config.h under a different name,
because running configure will destroy it.  If you have GNU autoconf
installed (the installation is very simple and source code is
available from prep.ai.mit.edu:/pub/gnu/autoconf-x.y.tar.gz), you
might try augmenting config.hin instead, then run autoconf and
configure.

Thus, on UNIX, installation normally consists of just two steps
(assuming a csh-compatible shell):

	setenv CC ...your favorite C or C++ compiler...
	./configure && make all check install

If you like, add OPT='your favorite optimization flags' to the make
command; by default, only -g (debug) is assumed.  If your compiler
won't accept -g with other optimization levels, then set CFLAGS
instead of OPT on the command line; be sure NOT to override any
non-optimizing flags in the CFLAGS set in the Makefile.

The GNU standard installation directories /usr/local/bin for binaries,
and /usr/local/man/man1 for manual pages are assumed.  The prefix
/usr/local can be overridden by providing an alternate definition on
the command line:

	make prefix=/some/other/path install

After installation, you can do
	make distclean
to restore the directories to their distribution state.  You should
also do this between builds for different architectures from the same
source tree; neglecting to do to will almost certainly lead to
failure, because the config.cache file created by configure will lead
to an incorrect config.h for the next build.


============
UNIX Systems
============

The code can be compiled with either C (K&R or ISO/ANSI Standard C) or
C++ compilers.  With some C++ compilers, it may be necessary to supply
additional switches for force the compiler to stay in C++ mode, rather
than reverting to C mode (e.g. on DEC Alpha OSF/1, you must do 
setenv CC "cxx -x cxx").

On UNIX systems, the only changes that you are likely to need in the
Makefile are the settings of CC and CFLAGS, and possibly, DEFINES, and
if you wish to do "make install", the settings of bindir, MANDIR, and
MANEXT.

If you are installing bibclean on a new system, you should definitely
run "make check" before installing it on your system.  For the target
test-bibtex-2, latex is needed.  For test-bibtex-2 and test-scribe-1,
bibtex is needed.  Sample output of "make check" from a UNIX system
is given below.

The code has been tested under more than 55 different C and C++
compilers, and is in regular use to maintain the TeX User Group
bibliography collection stored on ftp.math.utah.edu:/pub/tex/bib, as
well as several other local bibliographies.  These files total more
than 1.08M lines and 62K bibliography entries.  Some of these
bibliographies are mirrored to the Comprehensive TeX Archive Network
(CTAN) hosts.  Do "finger ctan@pip.shsu.edu" to find a CTAN site on
the Internet near you.

bibclean is also used for the BibNet Project, which collects
bibliographies in numerical analysis.  The master collection is
available on ftp.math.utah.edu:/pub/bibnet, and is mirrored from there
to netlib servers at AT&T and Oak Ridge National Laboratory.

If you port bibclean to a new system, please select maximal error and
warning messages in your compiler, to better uncover problems.  If you
find massive numbers of errors complaining about function and argument
type mismatches, it is likely that this can be remedied by suitable
modifications of config.h.  As C implementations move towards
conformance with the December 1989 ISO/ANSI C Language Standard, the C
language is a moving target that must be tracked by config.h, which is
why that file is normally automatically generated on UNIX systems by
the configure script.  With C compilers, you can safely ignore
complaints about implicit declaration of library functions; they are
caused by deficiencies in the vendor-provided header files.

If you have a C++ compiler, please try that as well.  This code has
been successfully compiled under at least 19 C++ compilers, and the
stricter type checking has uncovered problems that slipped past other
compilers.

These programs have been successfully built with C and C++ compilers
and tested on these systems for the 2.10.1 release:

	DECstation 5000		ULTRIX 4.2	cc, gcc, g++, lcc
	DEC Alpha		OSF/1 3.0, 3.2c	cc, c89, cxx, gcc, g++
	HP 9000/375		BSD 4.3		cc, CC
	HP 9000/735		HP-UX 9.0	cc, c89, CC, gcc, g++
	IBM RS/6000		AIX 3.2		cc, c89, xlC, gcc, g++
	Intel 486		Linux 1.3.15	gcc, g++
	MIPS RC6280		RISCos 2.1.1AC	cc
	NeXT 68040		Mach 3.0	cc, cc -ObjC, gcc, g++
	SGI 4D/210		IRIX 4.0.5c	cc, gcc, lcc
	SGI Indigo/2		IRIX 5.3	cc, CC, gcc, g++, lcc
	SGI Power Challenge	IRIX 6.0.1	cc, CC
	Sun SPARCstation	Solaris 2.3,2.4	cc, CC, gcc, g++, lcc
	Sun SPARCstation	SunOS 4.1.3	acc, cc, CC, gcc

Further details are given below.  Where builds have failed, it is
usually because of conflicts between system header files.

The author uses the build-all.sh script for these tests; it tries
builds with every known compiler on the development systems.  If your
UNIX system has other compilers that can be tested, please send their
full path names to the author.


==========
IBM PC DOS
==========


ChangeLog
2014-04-03 12:10:55-0600  Nelson H. F. Beebe  <beebe@math.utah.edu>

	* Release bibclean 2.16 after 15 months of extensive internal
	  use and testing.

2013-01-01 15:07:45-0700  Nelson H. F. Beebe  <beebe@math.utah.edu>

============================
Development of bibclean 2.16
============================

This is the first major update of bibclean in a decade, a period in
which it has successfully processed millions of BibTeX entries and
been an essential, reliable, and robust tool for creating high-quality
BibTeX data and databases. During that time, it has exhibited ZERO
bugs.

That decade of use, however, showed limitations in bibclean's cleanup
and correction of accents and title bracing, and the manual work
necessitated in further repairs of BibTeX entries generated from
publisher Web site data made it increasingly urgent to improve
bibclean so that a computer can do much more of the bibliographer's
drudge work.

During that decade, document URLs became common, and DOI (Digital
Object Identifier) and ISBN-13 book numbers were introduced.  In early
2013, 40% of the more than 740,000 entries in the TeX User Group and
BibNet Project bibliography archives contain URL values, 27% contain
DOI values, and 6% contain ISBN-13 values (but they appear in 82% of
Book/Proceedings entries).  Because ISBNs were only introduced in
1972, there will always be a substantial fraction of book-like entries
for which no ISBNs have been issued.

Only limited syntax checking on URLs and DOIs is possible, but because
a DOI is intended to be a unique document identifier, it should be
rare for there to be more than one DOI value, so this new version
checks for exceptions to that rule.  It is regrettable that the DOI
consortium at

	http://dx.doi.org/

did not learn from past mistakes, and require a simple but rigorous
syntax for all DOI values, and include error correction digits that
could be used to detect, and even repair, single and double character
errors.  All that can be sure about a DOI value is that it begins with
"10." followed by a decimal digit string identifying the publisher,
followed by a slash, followed by anything whatever.  In browser-usable
form, it begins "http://dx.doi.org/10."; that site then redirects it
to the current owner of the document (the original publisher might no
longer exist).

In 1997, this author developed bibsearch

	http://www.math.utah.edu/pub/bibsearch

to provide an interface to the MG (Managing Gigabytes) fast free-text
search engine.  Extensive experience with bibsearch shows that its
inability to restrict searches to document subfields is a major
limitation: entries can be found quickly (in just milliseconds).  As a
result, there are usually many entries found that are not of interest,
and the search often hides the sought-for golden needle in a file of
chaff.  Also, it is not possible to easily extract subsets of the
date, such as a list of unique authors or journals whose entries match
the search.

Consequently, in 2008, this author developed bibtosql and bibsql: see

	http://www.math.utah.edu/~beebe/talks/index.html#2009
	http://www.math.utah.edu/pub/bibsql

They provide a way to get clean (thanks to bibclean!) BibTeX data
into, and out of, three different SQL databases (MySQL, PostgreSQL,
and SQLite3).  That makes it possible to do precise subfield
searching, with SQL queries like this:

	select filename, year, label from bibtab
	    where (author like '%Knuth%')
	      and (title like '%TeX%')
	    order by filename, year, label;

It also allows implementing extensive sanity checks on BibTeX data
(e.g., find unexpected digits in author or editor values, probably
left there because of errors in conversion of publisher data that
superscripted author names to relate them to author address values).
Those sanity checks are now run automatically prior to Web publication
of newly-updated or newly-released BibTeX files at the sites

	http://www.math.utah.edu/pub/bibnet/
	http://www.math.utah.edu/pub/tex/bib/index-table.html

SQL searches also facilitate automated extraction of material from the
entire corpus for use in other BibTeX bibliographies.  Nightly cron
jobs produce temporary files containing possible updates to author-
and subject-specific bibliographies, making it much less likely that
new material is overlooked.  Cron jobs run several times daily update
the SQL databases with data from recently-changed BibTeX files.

An rcsdiff of the final code for bibclean version 2.16 shows about 530
replaced lines (out of about 15,260 original lines of code in version
2.15), and 2900 new lines.  Thus, version 2.16 is a SUBSTANTIAL update
that must receive extensive local use before it is released on the Web
at

	http://www.math.utah.edu/pub/bibclean

Here is a summary of the many changes for version 2.16:

	* Change license fields in comment preambles from "public
	  domain" to "GNU General Public License, version 2 or later".

	* Change copyright owner to include the Free Software
          Foundation.

	* Makefile.in: Change file order in cmp and diff commands in
	  check targets to conventional order "diff oldfile newfile".


option.c
#include <config.h>
#include "xlimits.h"
#include "xstdbool.h"
#include "xstdlib.h"
#include "xstring.h"

#include "ch.h"
#include "isbn.h"
#include "yesorno.h"
#include "match.h"
#include "typedefs.h"			/* must come AFTER match.h */

extern YESorNO	align_equals;		/* NO: left-adjust equals */
extern YESorNO	brace_protect;		/* YES: brace-protect title words */
extern YESorNO	check_values;		/* NO: suppress value checks */
extern YESorNO	debug_match_failures; 	/* YES: debug value pattern match failures */
extern YESorNO	delete_empty_values; 	/* YES: delete empty values */
extern YESorNO	fix_accents;		/* repair accent bracing? */
extern YESorNO	fix_braces;		/* normalize bracing? */
extern YESorNO	fix_font_changes;	/* brace {\em E. Coli}? */
extern YESorNO	fix_initials;		/* reformat A.U. Thor?	*/
extern YESorNO	fix_math;		/* reformat math mode? */
extern YESorNO	fix_names;		/* reformat Bach, P.D.Q? */
extern YESorNO	German_style;		/* YES: " inside braced string
					   value obeys german.sty style */
extern YESorNO	keep_linebreaks;	/* YES: keep linebreaks in values */
extern YESorNO	keep_parbreaks;		/* YES: keep parbreaks in values */
extern YESorNO	keep_preamble_spaces;	/* YES: keep spaces in @Preamble{} */
extern YESorNO	keep_spaces;		/* YES: keep spaces in values */
extern YESorNO	keep_string_spaces;	/* YES: keep spaces in @String{} */
extern long	max_width;
extern YESorNO	parbreaks;		/* NO: parbreaks forbidden */
					/* in strings and entries */
extern YESorNO	prettyprint;		/* NO: do lexical analysis */
extern YESorNO	print_ISBN_table;	/* YES: print ISBN table */
extern YESorNO	print_keyword_table;	/* YES: print keyword table */
extern YESorNO	print_patterns;		/* YES: print value patterns */
extern char	*program_name;		/* set to argv[0] */
extern YESorNO	read_initialization_files; /* -[no-]read-init-files sets */
extern YESorNO	remove_OPT_prefixes; 	/* YES: remove OPT prefix */
extern int	screen_lines;		/* kbopen() and out_lines() reset */
extern YESorNO	Scribe;			/* Scribe format input */
extern YESorNO	show_file_position; 	/* messages usually brief */
extern FILE	*stdlog;		/* usually stderr */
extern YESorNO	trace_file_opening;	/* -[no-]trace-file-opening sets */
extern YESorNO	warnings;		/* NO: suppress warnings */

static int	current_index;		/* argv[] index in do_args() */
static char	*current_option;	/* set by do_args() */
static char	*next_option;		/* set in do_args() */

void		do_args ARGS((int argc_, char *argv_[]));
void		do_preargs ARGS((int argc_, char *argv_[]));

bool		is_optionprefix ARGS((int c));

extern YESorNO	apply_function ARGS((const char *option_, OPTION_FUNCTION_ENTRY table_[]));
extern void	check_inodes ARGS((void));
extern void	do_initfile ARGS((const char *pathlist_,const char *name_));
extern void	do_keyword_file ARGS((const char *pathlist_,const char *name_));
extern void	out_lines ARGS((FILE *fpout_,const char *lines_[], YESorNO pausing_));
extern FILE	*tfopen ARGS((const char *filename_, const char *mode_));
extern void	warning ARGS((const char *msg_));

static void	opt_align_equals ARGS((void));
static void	opt_author ARGS((void));
static void	opt_brace_protect ARGS((void));
static void	opt_check_values ARGS((void));
static void	opt_copyleft ARGS((void));
static void	opt_debug_match_failures ARGS((void));
static void	opt_delete_empty_values ARGS((void));
static void	opt_error_log ARGS((void));
static void	opt_file_position ARGS((void));
static void	opt_fix_accents ARGS((void));
static void	opt_fix_braces ARGS((void));
static void	opt_fix_font_changes ARGS((void));
static void	opt_fix_initials ARGS((void));
static void	opt_fix_math ARGS((void));
static void	opt_fix_names ARGS((void));
static void	opt_help ARGS((void));
static void	opt_init_file ARGS((void));
static void	opt_ISBN_file ARGS((void));
static void	opt_German_style ARGS((void));
static void	opt_keep_linebreaks ARGS((void));
static void	opt_keep_parbreaks ARGS((void));
static void	opt_keep_preamble_spaces ARGS((void));
static void	opt_keep_spaces ARGS((void));
static void	opt_keep_string_spaces ARGS((void));
static void	opt_keyword_file ARGS((void));
static void	opt_max_width ARGS((void));
static void	opt_output_file ARGS((void));
static void	opt_parbreaks ARGS((void));
static void	opt_prettyprint ARGS((void));
static void	opt_print_ISBN_table ARGS((void));
static void	opt_print_keyword_table ARGS((void));
static void	opt_print_patterns ARGS((void));
static void	opt_quiet ARGS((void));
static void	opt_read_init_files ARGS((void));
static void	opt_remove_OPT_prefixes ARGS((void));
static void	opt_scribe ARGS((void));
static void	opt_trace_file_opening ARGS((void));
static void	opt_version ARGS((void));
static void	opt_warnings ARGS((void));
static void	usage ARGS((void));
static void	version ARGS((void));
static YESorNO	YESorNOarg ARGS((void));

static CONST char *help_lines[] =
{
    "\nUsage: ",
    (const char*)NULL,
    " [ -author ] [ -copyleft ] [ -copyright ]\n",
    "\t[ -error-log filename ] [ -help ] [ '-?'  ]\n",
    "\t[ -init-file filename ] [ -ISBN-file filename ]\n",
    "\t[ -keyword-file filename ] [ -max-width nnn ]\n",
    "\t[ -[no-]align-equals ] [ -[no-]brace-protect ]\n",
    "\t[ -[no-]check-values ] [ -[no-]debug-match-failures ]\n",
    "\t[ -[no-]delete-empty-values ] [ -[no-]file-position ]\n",
    "\t[ -[no-]fix-accents ] [ -[no-]fix-braces ]\n",
    "\t[ -[no-]fix-font-changes ] [ -[no-]fix-initials ]\n",
    "\t[ -[no-]fix-math ] [ -[no-]fix-names ] [ -[no-]German-style ]\n",
    "\t[ -[no-]keep-linebreaks ] [ -[no-]keep-parbreaks ]\n",
    "\t[ -[no-]keep-preamble-spaces ] [ -[no-]keep-spaces ]\n",
    "\t[ -[no-]keep-string-spaces ] [ -[no-]parbreaks ]\n",
    "\t[ -[no-]prettyprint ] [ -[no-]print-ISBN-table ]\n",
    "\t[ -[no-]print-keyword-table ] [ -[no-]print-patterns ]\n",
    "\t[ -[no-]quiet ] [ -[no-]read-init-files ]\n",
    "\t[ -[no-]remove-OPT-prefixes ] [ -[no-]scribe ]\n",
    "\t[ -[no-]trace-file-opening ] [ -[no-]warnings ]\n",
    "\t[ -output-file filename ] [ -version ]\n",
    "\t[ <infile or bibfile1 bibfile2 bibfile3 ...] >outfile\n",
    "\n",
#include "bibclean.h"
};


void
do_args(int argc, char *argv[])
{
    int k;				/* index into argv[] */
#define MSG_PREFIX	"Unrecognized option switch: "
#define MAX_OPTION_LENGTH	100
    char msg[sizeof(MSG_PREFIX) + MAX_OPTION_LENGTH + 1];
					/* for error messages */
    int nfiles;				/* number of files found in argv[] */

    static OPTION_FUNCTION_ENTRY options[] =
    {
	{"?",				1,	opt_help},
	{"align-equals",		2,	opt_align_equals},
	{"author",			2,	opt_author},
	{"brace-protect",		1,	opt_brace_protect },
	{"check-values",		1,	opt_check_values},
	{"copyleft",			2,	opt_copyleft },
	{"copyright",			2,	opt_copyleft },
	{"debug-match-failures",	3,	opt_debug_match_failures},
	{"delete-empty-values",		1,	opt_delete_empty_values},
	{"error-log",			1,	opt_error_log},
	{"file-position",		3,	opt_file_position},
	{"fix-accents",			5,	opt_fix_accents},
	{"fix-braces",			5,	opt_fix_braces},
	{"fix-font-changes",		5,	opt_fix_font_changes},
	{"fix-initials",		5,	opt_fix_initials},
	{"fix-math",			5,	opt_fix_math},
	{"fix-names",			5,	opt_fix_names},
	{"German-style",		1,	opt_German_style},
	{"help",			1,	opt_help},
	{"init-file",			2,	opt_init_file},
	{"ISBN-file",			2,	opt_ISBN_file},
	{"keep-linebreaks",		6,	opt_keep_linebreaks},
	{"keep-parbreaks",		7,	opt_keep_parbreaks},
	{"keep-preamble-spaces",	7,	opt_keep_preamble_spaces},
	{"keep-spaces",			7,	opt_keep_spaces},
	{"keep-string-spaces",		7,	opt_keep_string_spaces},
	{"keyword-file",		3,	opt_keyword_file},
	{"max-width",			1,	opt_max_width},
	{"no-align-equals",		4,	opt_align_equals},
	{"no-check-values",		4,	opt_check_values},
	{"no-debug-match-failures",	6,	opt_debug_match_failures},
	{"no-delete-empty-values",	6,	opt_delete_empty_values},
	{"no-file-position",		6,	opt_file_position},
	{"no-fix-accents",		8,	opt_fix_accents},
	{"no-fix-braces",		8,	opt_fix_braces},
	{"no-fix-font-changes",		8,	opt_fix_font_changes},
	{"no-fix-initials",		8,	opt_fix_initials},
	{"no-fix-math",			8,	opt_fix_math},
	{"no-fix-names",		8,	opt_fix_names},
	{"no-German-style",		4,	opt_German_style},
	{"no-keep-linebreaks",		9,	opt_keep_linebreaks},
	{"no-keep-parbreaks",		10,	opt_keep_parbreaks},
	{"no-keep-preamble-spaces",	10,	opt_keep_preamble_spaces},
	{"no-keep-spaces",		10,	opt_keep_spaces},
	{"no-keep-string-spaces",	10,	opt_keep_string_spaces},
	{"no-parbreaks",		5,	opt_parbreaks},
	{"no-prettyprint",		6,	opt_prettyprint},
	{"no-print-ISBN-table",		10,	opt_print_ISBN_table},
	{"no-print-keyword-table",	10,	opt_print_keyword_table},
	{"no-print-patterns",		10,	opt_print_patterns},
	{"no-quiet",			4,	opt_quiet},
	{"no-read-init-files",		6,	opt_read_init_files},
	{"no-remove-OPT-prefixes",	6,	opt_remove_OPT_prefixes},
	{"no-scribe",			4,	opt_scribe},
	{"no-trace-file-opening",	4,	opt_trace_file_opening},
	{"no-warnings",			4,	opt_warnings},
	{"output-file",			1,	opt_output_file},
	{"parbreaks",			2,	opt_parbreaks},
	{"prettyprint",			3,	opt_prettyprint},
	{"print-ISBN-table",		7,	opt_print_ISBN_table},
	{"print-keyword-table",		7,	opt_print_keyword_table},
	{"print-patterns",		7,	opt_print_patterns},
	{"quiet",			1,	opt_quiet},
	{"read-init-files",		3,	opt_read_init_files},
	{"remove-OPT-prefixes",		3,	opt_remove_OPT_prefixes},
	{"scribe",			1,	opt_scribe},
	{"trace-file-opening",		1,	opt_trace_file_opening},
	{"version",			1,	opt_version},
	{"warnings",			1,	opt_warnings},
	{(const char*)NULL,		0,	(void (*)(VOID))NULL},
    };

    for (nfiles = 1, k = 1; k < argc; ++k)
    {
	if ( (argv[k][1] != '\0') && is_optionprefix((int)argv[k][0]) )
	{				/* then process command-line switch */
	    current_index = k;		/* needed by opt_init_file() and */
	    next_option = argv[k+1];	/* opt_error_log() */
	    current_option = argv[k];	/* needed by YESorNOarg() */
	    if (is_optionprefix((int)current_option[1]))
		current_option++;	/* allow GNU/POSIX --option */
	    if (apply_function(current_option+1,options) == NO)
	    {
		(void)sprintf(msg, "%s%.*s", MSG_PREFIX, MAX_OPTION_LENGTH,
			      current_option);
		warning(msg);
		usage();
		exit(EXIT_FAILURE);
	    }
	    k = current_index;		/* some opt_xxx() functions update it */
	}
	else				/* save file names */
	    argv[nfiles++] = argv[k];	/* shuffle file names down */
    }
    argv[nfiles] = (char*)NULL;		/* terminate new argument list */
}


void
do_preargs(int argc, char *argv[])
{
    int k;

    static OPTION_FUNCTION_ENTRY options[] =
    {
	{"no-print-ISBN-table",		10,	opt_print_ISBN_table},
	{"no-print-patterns",		10,	opt_print_patterns},
	{"no-read-init-files",		6,	opt_read_init_files},
	{"no-trace-file-opening",	4,	opt_trace_file_opening},
	{"print-ISBN-table",		7,	opt_print_ISBN_table},
	{"print-patterns",		7,	opt_print_patterns},
	{"read-init-files",		3,	opt_read_init_files},
	{"trace-file-opening",		1,	opt_trace_file_opening},
	{(const char*)NULL,		0,	(void (*)(VOID))NULL},
    };

    for (k = 1; k < argc; ++k)
    {
	/* Do argument scan for options that must be known BEFORE
	initializations are attempted. */

	if ( (argv[k][1] != '\0') && is_optionprefix((int)argv[k][0]) )
	{				/* then process command-line switch */
	    current_index = k;
	    current_option = argv[k];
	    if (is_optionprefix((int)current_option[1]))
		current_option++;	/* allow GNU/POSIX --option */
	    next_option = argv[k+1];
	    (void)apply_function(current_option+1,options);
	}
    }
}


bool
is_optionprefix(int c)
{
    bool result = false;
    result = ((c) == (int)'-') ? true : false;
    return (result);
}


static void
opt_align_equals(VOID)
{
    align_equals = YESorNOarg();
}


static void
opt_author(VOID)
{
    static CONST char *author[] =
    {
	"Author:\n",
	"\tNelson H. F. Beebe\n",
	"\tUniversity of Utah\n",
	"\tDepartment of Mathematics, 110 LCB\n",
	"\t155 S 1400 E RM 233\n",
	"\tSalt Lake City, UT 84112-0090\n",
	"\tUSA\n",
	"\tTel: +1 801 581 5254\n",
	"\tFAX: +1 801 581 4148\n",
	"\tEmail: beebe@math.utah.edu, beebe@acm.org, beebe@computer.org (Internet)\n",
	(const char*)NULL,
    };

    out_lines(stdlog, author, NO);
    exit(EXIT_SUCCESS);
}

static void
opt_brace_protect(VOID)
{
    brace_protect = YESorNOarg();
}

static void
opt_check_values(VOID)
{
    check_values = YESorNOarg();
}


static void
opt_copyleft(VOID)
{
    static CONST char *copyleft[] =
    {
	"Copyright (C) 1990--2014 by Nelson H. F. Beebe and the Free Software Foundation, Inc.\n",
	"This is free software; see the source for copying conditions.  There is NO\n",
	"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",
	(const char*)NULL,
    };

    out_lines(stdlog, copyleft, NO);
    exit(EXIT_SUCCESS);
}


static void
opt_debug_match_failures(VOID)
{
    debug_match_failures = YESorNOarg();
}


static void
opt_delete_empty_values(VOID)
{
    delete_empty_values = YESorNOarg();
}


static void
opt_error_log(VOID)
{
    current_index++;

    if ((stdlog = tfopen(next_option,"w"), stdlog) == (FILE*)NULL)
    {
	(void)fprintf(stderr, "%s cannot open error log file [%s]",
		      WARNING_PREFIX, next_option);
	(void)fprintf(stderr, " -- using stderr instead\n");
	perror("perror() says");
	stdlog = stderr;
    }
    else
	check_inodes();				/* stdlog changed */
}


static void
opt_file_position(VOID)
{
    show_file_position = YESorNOarg();
}


static void
opt_fix_accents(VOID)
{
    fix_accents = YESorNOarg();
}


static void
opt_fix_braces(VOID)
{
    fix_braces = YESorNOarg();
}


static void
opt_fix_font_changes(VOID)
{
    fix_font_changes = YESorNOarg();
}


static void
opt_fix_initials(VOID)
{
    fix_initials = YESorNOarg();
}


static void
opt_fix_math(VOID)
{
    fix_math = YESorNOarg();
}


static void
opt_fix_names(VOID)
{
    fix_names = YESorNOarg();
}


static void
opt_German_style(VOID)
{
    German_style = YESorNOarg();
}


static void
opt_help(VOID)
{
    help_lines[1] = program_name;	/* cannot have this in initializer */
    out_lines(stdlog, help_lines, (screen_lines > 0) ? YES : NO);
    exit(EXIT_SUCCESS);
}


static void
opt_init_file(VOID)
{
    current_index++;
    do_initfile((const char*)NULL,next_option);
}


static void
opt_ISBN_file(VOID)
{
    current_index++;
    do_ISBN_file((const char*)NULL,next_option);
}


static void
opt_keep_linebreaks(VOID)
{
    keep_linebreaks = YESorNOarg();
}


static void
opt_keep_parbreaks(VOID)
{
    keep_parbreaks = YESorNOarg();
}


static void
opt_keep_preamble_spaces(VOID)
{
    keep_preamble_spaces = YESorNOarg();
}


static void
opt_keep_spaces(VOID)
{
    keep_spaces = YESorNOarg();
}


static void
opt_keep_string_spaces(VOID)
{
    keep_string_spaces = YESorNOarg();
}


static void
opt_keyword_file(VOID)
{
    current_index++;
    do_keyword_file((const char*)NULL,next_option);
}


static void
opt_max_width(VOID)
{
    current_index++;
    max_width = strtol(next_option,(char**)NULL,0);
    if (max_width <= 0L)		/* width <= 0 means unlimited width */
	max_width = LONG_MAX;
}


static void
opt_output_file(VOID)
{
    current_index++;

    if (freopen(next_option, "w", stdout) != stdout)
    {
	(void)fprintf(stderr, "Cannot open file [%s] for output: job terminated!\n", next_option);
	exit(EXIT_FAILURE);
    }
}


static void
opt_parbreaks(VOID)
{
    parbreaks = YESorNOarg();
}


static void
opt_prettyprint(VOID)
{
    prettyprint = YESorNOarg();
}


static void
opt_print_ISBN_table(VOID)
{
    print_ISBN_table = YESorNOarg();
}


static void
opt_print_keyword_table(VOID)
{
    print_keyword_table = YESorNOarg();
}


static void
opt_print_patterns(VOID)
{
    print_patterns = YESorNOarg();
}


static void
opt_quiet(VOID)
{
    warnings = (YESorNOarg() == YES) ? NO : YES;
}


static void
opt_read_init_files(VOID)
{
    read_initialization_files = YESorNOarg();
}


static void
opt_remove_OPT_prefixes(VOID)
{
    remove_OPT_prefixes = YESorNOarg();
}


static void
opt_scribe(VOID)
{
    Scribe = YESorNOarg();
}


static void
opt_trace_file_opening(VOID)
{
    trace_file_opening = YESorNOarg();
}


static void
opt_version(VOID)
{
    version();
    exit(EXIT_SUCCESS);
}


static void
opt_warnings(VOID)
{
    warnings = YESorNOarg();
}


static void
usage(VOID)
{
    version();
    help_lines[1] = program_name;	/* cannot have this in initializer */
    out_lines(stdlog, help_lines, NO);
}


static void
version(VOID)
{
    static CONST char *version_string[] =
    {
	PACKAGE_NAME, " Version ", PACKAGE_VERSION, " [", PACKAGE_DATE, "]",
	"\n",
	"\n",

#if defined(HOST) || defined(USER) || defined(__DATE__) || defined(__TIME__)
	"Compiled",

#if defined(USER)
	" by <", USER,

#if defined(HOST)
	"@", HOST,
#endif /* defined(HOST) */

	">",
#endif /* defined(USER) */

#if defined(__DATE__)
	" on ", __DATE__,
#endif /* defined(__DATE__) */

#if defined(__TIME__)
	" ", __TIME__,
#endif /* defined(__TIME__) */

#if defined(HAVE_PATTERNS)
	"\nwith native pattern matching",
#endif /* defined(HAVE_PATTERNS) */

#if defined(HAVE_RECOMP) || defined(HAVE_REGEXP)
	"\nwith regular-expression pattern matching",
#endif /* defined(HAVE_RECOMP) || defined(HAVE_REGEXP) */

#if defined(HAVE_OLDCODE)
	"\nwith old matching code",
#endif /* defined(HAVE_OLDCODE) */

	"\n",
#endif /* defined(HOST)||defined(USER)||defined(__DATE__)||defined(__TIME__) */
	"E-mail bug reports to ", PACKAGE_BUGREPORT, "\n",

	"\n",
	"Copyright (C) 1990--2014 by Nelson H. F. Beebe and the Free Software Foundation, Inc.\n",
	"This is free software; see the source for copying conditions.  There is NO\n",
	"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",

	(const char*)NULL,
    };

    out_lines(stdlog, version_string, NO);
}


static YESorNO
YESorNOarg(VOID)
{
    return ((strnicmp(current_option+1,"no-",3) == 0) ? NO : YES);
}
